processing/src/main/java/org/apache/druid/query/dimension/ListFilteredDimensionSpec.java

```diff
@@ -92,15 +92,17 @@ public static IdMapping buildAllowListIdMapping(
       IndexedGetter<String> fn
   )
   {
-    final IdMapping.Builder builder = IdMapping.Builder.ofCardinality(values.size());
+    IdMapping.Builder builder;
     if (idLookup != null) {
+      builder = IdMapping.Builder.ofCardinality(values.size());
       for (String value : values) {
         int i = idLookup.lookupId(value);
         if (i >= 0) {
           builder.addMapping(i);
         }
       }
     } else {
+      builder = IdMapping.Builder.ofCardinality(cardinality);
```

Comment:
Hmm, I don't think we probably want to do this like this, rather, I think we wan to switch [`IdMapping.Builder` to use a growable collection like `IntList` instead of `int[]` for the `reverseMapping`](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/segment/IdMapping.java#L102) to handle this case where the ids are a many to one mapping because of something like a lookup extraction fn being used. The reason is because for something like `MV_FILTER_ONLY`, the cardinality of the column might be like 1 million, while the cardinality of the values supplied to the function might just be a few values, so allocating a whole `int[]` for the complete cardinality of the column seems pretty dramatic.

We could potentially optimize this in some manner by checking for extractionFn, and [if the extractionFn extraction type is one to one, and only do this path if it is many to one, but I don't think its really needed](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/query/extraction/ExtractionFn.java#L118).

This matters a lot less for a deny list, which likely will still have a cardinality close to the base column.

Reply:
I didn't wanted to change that class - but it makes sense to do so.

with auto-growing available this second branch doesn't cause any issues; so I've undone the changes to this part.

btw: I'm wondering in case it reports 1-to-1 - wouldn't that still map unknown values to `null` ? ...so if there is a `null` in the input `values` it could still have more matches - but anyway this will work correctly with the current approach

Reply:
hmm, actually is that right? Thinking a bit more about this, maybe I'm wrong and reverse mapping actually needs to be a list of ints associated with each value if there is a many to one mapping.. which introduces a lot of extra complexity in places (i've also found some bugs in the index supplier implementation of this virtual column). Let me think a bit more about this...

Reply:
ok, ignore all of this, we talked about this offline, but to capture that discussion here: I think the _real_ right thing to do here is have `ListFilteredVirtualColumn` and `ListFilteredDimensionSpec` force use of `PredicateFilteredDimensionSelector` whenever the extractionFn is present on the `delegate` `DimensionSpec`, for both allow or deny lists.

Additionally `ListFilteredVirtualColumn` also should return `NoIndexesColumnIndexSupplier.getInstance()` whenever `getIndexSupplier` is called and the `delegate` has an `extractionFn`.

The reason is that use of `IdMapping` requires a 1:1 relation between the filtered dimension and underlying dimension, and that isn't really possible to do with a many to one lookup.